File: conversions/nanoid/elide-nanoid.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-nanoid (Compact ID Generator)

Demonstrates calling the TypeScript nanoid implementation from Python
for consistent compact URL-safe IDs across all services.

Benefits:
- One ID generator shared across TypeScript and Python
- Consistent ID format across all microservices
- No Python uuid library needed
- Shorter, URL-safe IDs (21 chars vs 36 for UUID)
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# from elide import require
# nanoid_module = require('./elide-nanoid.ts')

print("=== Python Calling TypeScript Nanoid ===\n")

# Example 1: Generate compact IDs
# id1 = nanoid_module.nanoid()
# print(f"ID 1: {id1}")  # e.g., "V1StGXR8_Z5jdHi6B-myT"
#
# id2 = nanoid_module.nanoid()
# print(f"ID 2: {id2}")  # e.g., "Uakgb_J5m9g-0JDMbcJqL"

# Example 2: Custom sizes
# short_id = nanoid_module.nanoid(10)
# print(f"Short ID (10): {short_id}")  # e.g., "IRFa-VaY2b"
#
# long_id = nanoid_module.nanoid(32)
# print(f"Long ID (32): {long_id}")  # e.g., "aWY7IFgXqVFRq_PqL5kTj7Pqb5hY3qT"

# Example 3: Custom alphabet (numbers only)
# numbers_gen = nanoid_module.customAlphabet(nanoid_module.alphabets.numbers)
# numeric_id = numbers_gen(10)
# print(f"Numeric ID: {numeric_id}")  # e.g., "7482910536"

# Example 4: Django models with compact IDs
# from django.db import models
#
# class UrlShortener(models.Model):
#     short_code = models.CharField(max_length=8, unique=True)
#     long_url = models.URLField()
#     created_at = models.DateTimeField(auto_now_add=True)
#
#     @staticmethod
#     def generate_short_code():
#         # Use nanoid for consistent short codes
#         alphanum = nanoid_module.customAlphabet(
#             nanoid_module.alphabets.alphanumeric, 8
#         )
#         return alphanum()
#
# # Create shortened URL
# shortener = UrlShortener(
#     short_code=UrlShortener.generate_short_code(),
#     long_url='https://example.com/very/long/url'
# )
# print(f"Short URL: https://short.ly/{shortener.short_code}")

# Example 5: Flask API with nanoid
# from flask import Flask, jsonify
# app = Flask(__name__)
#
# @app.route('/api/generate-id')
# def generate_id():
#     # Same ID format as Node.js API!
#     new_id = nanoid_module.nanoid(16)
#     return jsonify({"id": new_id})
#
# @app.route('/api/short-url', methods=['POST'])
# def create_short_url():
#     short_code_gen = nanoid_module.customAlphabet(
#         nanoid_module.alphabets.alphanumeric, 8
#     )
#     short_code = short_code_gen()
#     return jsonify({
#         "short_url": f"https://short.ly/{short_code}",
#         "short_code": short_code
#     })

# Example 6: Database IDs (shorter than UUID)
# import sqlite3
#
# def create_user(username, email):
#     user_id = f"user_{nanoid_module.nanoid(16)}"  # e.g., "user_V1StGXR8_Z5jdHi"
#     # Insert into database
#     # INSERT INTO users (id, username, email) VALUES (?, ?, ?)
#     return user_id

print("Real-world use case:")
print("- Python API generates short URL codes")
print("- Uses same nanoid as Node.js service")
print("- Guarantees identical ID format across entire stack")
print()

print("Example: URL Shortener")
print("┌─────────────────────────────────────┐")
print("│   Elide Nanoid (TypeScript)        │")
print("│   conversions/nanoid/elide-nanoid.ts│")
print("└─────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  API   │          │  API   │")
print("    └────────┘          └────────┘")
print("     ID: 'aB3x9K1z'     ID: 'aB3x9K1z'")
print("     = 8 chars           = 8 chars")
print("     ✓ Consistent URL-safe IDs!")
print()

print("Problem Solved:")
print("Before: Node.js nanoid + Python uuid = different ID formats")
print("After: One Elide implementation = identical compact IDs")
print()

print("Configuration Example:")
print("  # Both Node.js and Python generate same format:")
print("  short_code = nanoid(8)  # 'aB3x9K1z' in both languages")
print("  url = f'https://short.ly/{short_code}'")
print()

print("Benefits:")
print("  ✓ 60% smaller than UUID (21 vs 36 chars)")
print("  ✓ URL-safe (no special encoding needed)")
print("  ✓ Collision-resistant")
print("  ✓ Consistent across all services")

File: converted/utilities/nanoid/elide-nanoid.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-nanoid (Compact ID Generator)
#
# Demonstrates calling the TypeScript nanoid implementation from Ruby
# for consistent compact URL-safe IDs across microservices.

puts "=== Ruby Calling TypeScript Nanoid ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# nanoid_module = Elide.require('./elide-nanoid.ts')

# Example 1: Generate compact IDs
# id1 = nanoid_module.nanoid()
# puts "ID 1: #{id1}"  # e.g., "V1StGXR8_Z5jdHi6B-myT"

# Example 2: Custom sizes for URL shortener
# short_code_gen = nanoid_module.customAlphabet(
#   nanoid_module.alphabets[:alphanumeric], 8
# )
# short_code = short_code_gen.call()
# puts "Short code: #{short_code}"  # e.g., "aB3x9K1z"

# Example 3: Rails model with nanoid
# class ShortenedUrl < ApplicationRecord
#   before_create :generate_short_code
#
#   private
#
#   def generate_short_code
#     nanoid = Elide.require('./elide-nanoid.ts')
#     gen = nanoid.customAlphabet(nanoid.alphabets[:alphanumeric], 8)
#     self.short_code = gen.call()
#   end
# end
#
# # Create shortened URL
# url = ShortenedUrl.create(
#   long_url: 'https://example.com/very/long/url'
# )
# puts "Short URL: https://short.ly/#{url.short_code}"

# Example 4: Sidekiq job with unique IDs
# class ProcessJob
#   include Sidekiq::Worker
#
#   def perform(data)
#     nanoid = Elide.require('./elide-nanoid.ts')
#     job_id = nanoid.nanoid(16)
#     logger.info "Processing job: #{job_id}"
#     # Process with unique ID, same format as Node.js!
#   end
# end

# Example 5: Sinatra API
# require 'sinatra'
# NANOID = Elide.require('./elide-nanoid.ts')
#
# get '/api/generate-id' do
#   content_type :json
#   { id: NANOID.nanoid(16) }.to_json
# end
#
# post '/api/short-url' do
#   gen = NANOID.customAlphabet(NANOID.alphabets[:alphanumeric], 8)
#   short_code = gen.call()
#   content_type :json
#   {
#     short_url: "https://short.ly/#{short_code}",
#     short_code: short_code
#   }.to_json
# end

# Example 6: Database IDs (shorter than UUID)
# class User < ApplicationRecord
#   before_create :generate_compact_id
#
#   private
#
#   def generate_compact_id
#     nanoid = Elide.require('./elide-nanoid.ts')
#     self.id = "user_#{nanoid.nanoid(16)}"  # e.g., "user_V1StGXR8_Z5jdHi"
#   end
# end

# Example 7: Session tokens
# class SessionManager
#   def self.create_token
#     nanoid = Elide.require('./elide-nanoid.ts')
#     nanoid.nanoid(32)  # Secure 32-char token
#   end
# end

puts "Real-world use case:"
puts "- Ruby app generates short URL codes"
puts "- Uses same nanoid as Node.js service"
puts "- Guarantees identical ID format across entire platform"
puts ""

puts "Example: URL Shortener Service"
puts "┌─────────────────────────────────────┐"
puts "│   Elide Nanoid (TypeScript)        │"
puts "│   conversions/nanoid/elide-nanoid.ts│"
puts "└─────────────────────────────────────┘"
puts "         ↓           ↓"
puts "    ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Ruby  │"
puts "    │  API   │  │  Rails │"
puts "    └────────┘  └────────┘"
puts "     short.ly/aB3x9K1z"
puts "     ✓ Consistent short codes!"
puts ""

puts "Problem Solved:"
puts "Before: Different ID generators = inconsistent formats"
puts "After: One Elide implementation = perfect synchronization"
puts ""

puts "Benefits:"
puts "  ✓ 60% smaller than UUID (21 vs 36 chars)"
puts "  ✓ URL-safe (no special encoding needed)"
puts "  ✓ Collision-resistant"
puts "  ✓ Same format in Node.js, Python, Ruby, Java"
